### ISSUE TITLE
How to render to Android's SurfaceTexture

### ISSUE THREAD

**Issue by lattice0** [glium/glium#2009]
Looking at a pure-rust android example https://github.com/kmcallister/azurescens/blob/master/src/main.rs#L153 we see that we need to pass a Facade, which is a display built from a WindowBuilder. What if I want to NOT use any Android windowing and instead render to an android's `SurfaceTexture`? Is it possible to get a display from a `SurfaceTexture`?

**Comment by lattice0** [comment]
Looks like I need to implement a new Backend for SurfaceTexture: https://github.com/glium/glium/blob/master/src/backend/mod.rs#L36

Looking at the capabilities of SurfaceTexture https://developer.android.com/ndk/reference/group/surface-texture#summary 

I think that

```
    /// Makes the OpenGL context the current context in the current thread.
    unsafe fn make_current(&self);
```

maps to

[ASurfaceTexture_attachToGLContext](https://developer.android.com/ndk/reference/group/surface-texture#group___surface_texture_1ga0a204415e7f93b6b2a09cd4d71d3bb39)([ASurfaceTexture](https://developer.android.com/ndk/reference/group/surface-texture#group___surface_texture_1ga5c6696f8d5335d9b71fbeff5f8265c7c) *st, uint32_t texName)


and 

```
 /// Returns true if the OpenGL context is the current one in the thread.
    fn is_current(&self) -> bool;
```

can be simulated somehow based on each SurfaceTexture being marked as active or not when this is called. 

Maybe 

`fn get_framebuffer_dimensions(&self) -> (u32, u32);`

is the size of the SurfaceTexture which is defined at creation so I can use that.

I just don't know what to do with

    ```
/// Swaps buffers at the end of a frame.
    fn swap_buffers(&self) -> Result<(), SwapBuffersError>;
```
and maybe
```

    /// Returns the address of an OpenGL function.
    ///
    /// Supposes that the context has been made current before this function is called.
    unsafe fn get_proc_address(&self, symbol: &str) -> *const c_void;
```

**Comment by lattice0** [comment]
I think that

[ASurfaceTexture_updateTexImage](https://developer.android.com/ndk/reference/group/surface-texture#group___surface_texture_1ga254103ee3bedcde3433c756a70967445)([ASurfaceTexture](https://developer.android.com/ndk/reference/group/surface-texture#group___surface_texture_1ga5c6696f8d5335d9b71fbeff5f8265c7c) *st)

is also quite important but I don't see what would the Backend map it too

**Comment by est31** [comment]
PRs welcome!

**Comment by lattice0** [comment]
@est31 I'm definetly going to make a PR, I already have one bringing SurfaceTexture to Rust: https://github.com/rust-windowing/android-ndk-rs/pull/267

But I'm not an expert in OpenGL, so I need to know what to do with `    fn swap_buffers(&self) -> Result<(), SwapBuffersError>;` and [ASurfaceTexture_updateTexImage](https://developer.android.com/ndk/reference/group/surface-texture#group___surface_texture_1ga254103ee3bedcde3433c756a70967445)([ASurfaceTexture](https://developer.android.com/ndk/reference/group/surface-texture#group___surface_texture_1ga5c6696f8d5335d9b71fbeff5f8265c7c) *st) and [ASurfaceTexture_detachFromGLContext](https://developer.android.com/ndk/reference/group/surface-texture#group___surface_texture_1ga4cfaf4574f08cc7b0c427776879315aa)([ASurfaceTexture](https://developer.android.com/ndk/reference/group/surface-texture#group___surface_texture_1ga5c6696f8d5335d9b71fbeff5f8265c7c) *st) as these and the Backend do not go together

**Comment by lattice0** [comment]
Is it ok to use this EGL wrapper? https://github.com/timothee-haudebourg/khronos-egl

**Comment by est31** [comment]
Why does glutin not work? They use egl themselves. There, the bindings are auto generated during build time. Can your feature maybe be added to a platform specific glutin extension?

**Comment by lattice0** [comment]
because glutin is for rendering to a window, I need to render to a SurfaceTexture. I made a PR that adds SurfaceTexture to glium as a backend (and also one to android-ndk-rs https://github.com/rust-windowing/android-ndk-rs/pull/267#pullrequestreview-988427860), but when I tried to glue `attachToGlContext` to `make_current` I get `attachToContext: invalid current EGLDisplay` and then `Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 8548 (rfacetexturepoc), pid 8548 (rfacetexturepoc)` then looking at this error I found https://android.googlesource.com/platform/frameworks/native/+/a0db308/libs/gui/SurfaceTexture.cpp#387 which suggests that there's no EGL display, so I had to create one using an EGL rust wrapper.

I wish I could do this with glium only, I just need the OpenGL functionality into SurfaceTexture

If you want to see my very crazy try: https://github.com/lattice0/glium_surface_texture_poc/blob/1e8f669f9dee7d6ef8dd6589fac31754f9c0aea9/app/src/main/rust/surface_texture_glium/src/lib.rs this is where I try to create the EGL display

**Comment by est31** [comment]
glutin is not just for rendering to windows, but for OpenGL context creation in general. They explicitly support headless use cases. Ask them whether they accept a contribution. If they don't want it, or want it done in a way that doesn't fit your use cases, I'm open to using EGL directly in glium, but then it should still not go via the egl crate, but via extending glium's gl-generator use in build.rs. That dependency already exists and it's also what glutin is doing.

**Comment by MarijnS95** [comment]
I think this can just go into `glutin`, I'm already working on bringing the Android support back to life with all the extra surface/window lifetime management, sneaking a couple functions in there that accept an `ndk` `SurfaceTexture` and/or `NativeWindow` should be no big deal.

**Comment by lattice0** [comment]
For me headless seems like rendering to a buffer not backed by anything, but here I need it backed by a SurfaceTexture. 

I think this should be the right choice: https://github.com/lattice0/glutin/blob/master/glutin/src/api/egl/mod.rs, but I think my glium implementation for SurfaceTexture as a backend might still be needed? https://github.com/lattice0/glium/blob/8d21ea1e1c72138f5f878393736e7fcfdfae6716/src/backend/surface_texture/mod.rs

For what I understand, I need to create a `NativeWindow` from a `Surface`, which is created from a `SurfaceTexture`. The word "window" here makes no sense to me as there's no window, but anyways, a window is something that can be used to create an EGLSurface: `EGL::surface = eglCreateWindowSurface(EGL::display, EGL::config, ::window, nullptr);` which is what I need, so I can call make current into it, etc. I don't know the difference for a display, but looks like an egl surface is sufficient https://github.com/StenSoft/Android-Surfacing/blob/master/surfacing/cpp/MainActivity.cpp#L208

@MarijnS95 I'd like to help and not just sit and wait for your integration

**Comment by est31** [comment]
> @MarijnS95 I'd like to help and not just sit and wait for your integration

FTR one can formulate such a statement more nicely. Being impatient is fair, being rude is not. Nobody is paid here for their work, it's all hobbyists, and especially not by you. I would strongly suggest you apologize to @MarijnS95 for that statement. You will likely need their collaboration.

I also think that this discussion is best continued in a dedicated glutin issue. Therefore, I'm closing this.

**Comment by lattice0** [comment]
@est31 in my head this comment was the opposite of rude, I was trying to say that I want to help and not just sit and wait for him to do all the job. Sorry if it sounded rude, @MarijnS95. I want to actually help :)

**Comment by MarijnS95** [comment]
FTR I don't use OpenGL, all the contributions and replies you saw from me related to these issues and "finally making `glutin` work on Android" stem from a desperate attempt to close a whole bunch of issues and PRs that need to be "babysitted through" to get them cleaned and merged. Of course not everyone sits on the same experience/knowledge level straight away and that's totally understandable/fine, but at this point I think it's more efficient if I implement this on my own and let you test the PR(s) when it is done! Thanks for your patience!

Thanks @est31, apologies for the noise here! We'll move over to `glutin` where a dedicated/duplicated issue already exists.

**Comment by est31** [comment]
> I was trying to say that I want to help and not just sit and wait for him to do all the job.

Oh I see, I guess it's one of those cases where multiple interpretations exist. Sorry for misunderstanding then. It's all good.

> We'll move over to glutin where a dedicated/duplicated issue already exists.

I suppose that's https://github.com/rust-windowing/glutin/issues/1324 ?

**Comment by MarijnS95** [comment]
Correct. @lattice0 spawned quite a few issues/PRs around this but I think that's the canonical one when it comes to implementing this support on `glutin` specifically. The rest is just making things compile/work on Android (again) in the first place...

**Comment by MarijnS95** [comment]
For future reference:

https://github.com/MarijnS95/AndroidNativeSurface/compare/glium

This works through some quick hacks in `glutin` to accept a generic `HasRawWindowHanlde` (cc https://github.com/rust-windowing/android-ndk-rs/pull/274) and in turn any Android `Surface` (from `SurfaceView`) or `SurfaceTexture` (from `TextureView` or Kotlin).

**Comment by lattice0** [comment]
@MarijnS95 thank you **very** much, I still couldn't test but gonna do it soon. Wouldn't the window handle be to the entire android app window? Cause I'm gonna use 20 surface textures. Wouldn't it still make sense to implement SurfaceTexture as a Glium backend?

**Comment by MarijnS95** [comment]
@lattice0 I may have explained my presumption before, but here goes.

As far as I can tell a `NativeWindow` was never nothing more than a C representation for a Java `Surface`. At first, when the NDK was initially crafted, this is totally fine. The user has a window to which they can render, and is oblivious to what it represents in the background.

Then came the idea, or rather extra bindings, to have an arbitrary `Surface` on a `View` residing somewhere in Java's `View` hierarchy, which is pretty much what the `NativeWindow` is that you get through their `onNativeWindowCreated` callback; just that you can now create it yourself (disclaimer: I don't know if this is historically accurate, perhaps these bindings existed from the get-go). It just seems that the NDK developers didn't get to, or couldn't, rename the type anymore, even though it is just wrapping the `Surface` and nothing else.

Disclaimer, just to be extra sure: this is all mere speculation.

The only important thing to take away is that: _no_, `NativeWindow` does _not_ represent the entire window, ever. It only represents a `Surface` in a Java `View` hierarchy which _may_ be configured to take up the entire window (with or without the status/navigation bar drawing over it by default...). The naming is odd indeed, but this is how it works.

Perhaps an image says more than a thousand words:

![image](https://user-images.githubusercontent.com/2325264/171288692-40ad4efb-49ae-443a-8d5d-f62f7d3c9698.png)

**Comment by MarijnS95** [comment]
> Wouldn't it still make sense to implement SurfaceTexture as a Glium backend?

I haven't played enough with the difference between the two to say for sure. I must've mentioned elsewhere that these examples work without binding a context/texture to the `SurfaceTexture` nor calling `update_tex_image()`, so I presume this is the way to use it (but some animated GL and persisting the context is a great way to test).

**Comment by lattice0** [comment]
Very nice! I started with glutin on desktop then realized that I needed a SurfaceTexture, then I moved to glium only, because glium is just a nicer way to call OpenGL functions, and I had already done that before: rendering to a SurfaceTexture with OpenGL calls. That's why I was implementing a Backend for SurfacceTexture, because glutin both seemed like an overkill and unnecessary. But then I realized I'd have to have EGL calls on the backend, for which glutin would be needed, or another EGL wrapper but they wouldn't accept another one
